waku/waku_rendezvous/protocol.nim

```diff
@@ -1,117 +1,157 @@
 {.push raises: [].}
 
 import
-  std/[sugar, options],
+  std/[sugar, options, sequtils, algorithm, tables],
   results,
   chronos,
   chronicles,
-  metrics,
+  stew/byteutils,
   libp2p/protocols/rendezvous,
+  libp2p/protocols/rendezvous/protobuf,
+  libp2p/discovery/discoverymngr,
+  libp2p/utils/semaphore,
+  libp2p/utils/offsettedseq,
+  libp2p/crypto/curve25519,
   libp2p/switch,
   libp2p/utility
 
+import metrics except collect
+
 import
   ../node/peer_manager,
   ../common/callbacks,
   ../waku_enr/capabilities,
   ../waku_core/peers,
+  ../waku_core/codecs,
   ../waku_core/topics,
   ../waku_core/topics/pubsub_topic,
-  ./common
+  ./common,
+  ./waku_peer_record
+
+# Explicitly make the items iterator visible in this module's scope
+# This is needed for generic instantiation of rendezvous methods
+iterator items*[T](o: OffsettedSeq[T]): T =
+  for item in offsettedseq.items(o):
+    yield item
 
 logScope:
   topics = "waku rendezvous"
 
 declarePublicCounter rendezvousPeerFoundTotal,
   "total number of peers found via rendezvous"
 
-type WakuRendezVous* = ref object
-  rendezvous: Rendezvous
+type WakuRendezVous* = ref object of GenericRendezVous[WakuPeerRecord]
   peerManager: PeerManager
   clusterId: uint16
   getShards: GetShards
   getCapabilities: GetCapabilities
+  getPeerRecord: GetWakuPeerRecord
 
   registrationInterval: timer.Duration
   periodicRegistrationFut: Future[void]
 
   requestInterval: timer.Duration
   periodicRequestFut: Future[void]
-
-proc batchAdvertise*(
+  clientOnlyMode: bool
```

Comment:
Not sure if I'm missing some subtlety, but I find it a bit weird that `client` is a state on `WakuRendezVous`. Why not split into a `...client` type (and module) and service/protocol module as we do for other protocols? I assume there might be too much shared state to make it worth the effort, but I cannot see any fundamental reason why not in the `discover` and `advertise` below.